### ISSUE TITLE
AWS SDK V3 Requests Failing on Alternate Calls

### ISSUE THREAD

**Issue by ElusiveMind** [minio/minio#7728]
## Expected Behavior
When I run an API call, it should return the result of that call.

## Current Behavior
The PHP API call to Minio fails on alternating calls.  The following is when I am trying to obtain a list of all objects in a bucket

## Steps to Reproduce (for bugs)
`[serviceadmin@CentOS03 minio]$ php fsis.php`
`drupal-data-2019-05-16.sql.gz`
`[serviceadmin@CentOS03 minio]$ php fsis.php `
`Error executing "ListObjects" on "http://xxxxx/drupal-data?encoding-type=url"; AWS HTTP error: Client error: GET http://xxxxx/drupal-data?encoding-type=url resulted in a 404 Not Found response NotFound (client): 404 Not Found - `
`[serviceadmin@CentOS03 minio]$ `

## Your Environment
1. Using latest docker version of containerized Minio
2. Proxied through Nginx
3. Running on CentOS v7
4. Code is not in a repository - using examples from S3 documentation

**Comment by nitisht** [comment]
Can you set `proxy_http_version 1.1` in your Nginx config and try again?

**Comment by ElusiveMind** [comment]
I just did and am getting the same result. Here is my server configuration

 ` server {`
  `  listen 80;`
  `  client_max_body_size 5000M;`
  `  server_name xxxx;`
  `  location / {`
  `    proxy_pass http://127.0.0.1:8201;`
  `    proxy_set_header Host $host;`
  `    proxy_set_header X-Real-IP $remote_addr;`
  `    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;`
  `    proxy_set_header X-Forwarded-Proto $scheme;`
  `    proxy_connect_timeout       3600;`
  `    proxy_send_timeout          3600;`
  `    proxy_read_timeout          3600;`
  `    send_timeout                3600;`
  `    proxy_http_version          1.1;`
  `  }`
  `}`

**Comment by nitisht** [comment]
can you share the php code for us to take a look ? Also, can you confirm if `mc ls` works consistently?

**Comment by ElusiveMind** [comment]
I am going to be looking into this after work today and will report back

**Comment by ElusiveMind** [comment]
I have attached the script file minus authentication pieces.

Keep in mind the domain goes to an Nginx reverse proxy which sends it to an Apache/PHP docker on port 9000.

[alternating-error.txt](https://github.com/minio/minio/files/3243870/alternating-error.txt)

**Comment by nitisht** [comment]
> Also, can you confirm if mc ls works consistently?

Can you check this as well?

**Comment by ElusiveMind** [comment]
Weird.... I thought it was.. Then...

NuMacOS:Desktop mbagnall$ ./mc ls fsis
[2019-06-02 09:13:42 EDT]      0B crawler-data/
[2019-06-02 07:53:12 EDT]      0B databases/
[2019-06-02 09:00:40 EDT]      0B drupal-data/
[2019-06-02 07:52:31 EDT]      0B drupal-files/
[2019-06-02 07:52:09 EDT]      0B private-files/
[2019-06-02 09:09:46 EDT]      0B solr-cores/
NuMacOS:Desktop mbagnall$ ./mc ls fsis
[2019-06-02 09:13:42 EDT]      0B crawler-data/
[2019-06-02 07:53:12 EDT]      0B databases/
[2019-06-02 09:00:40 EDT]      0B drupal-data/
[2019-06-02 07:52:31 EDT]      0B drupal-files/
[2019-06-02 07:52:09 EDT]      0B private-files/
[2019-06-02 09:09:46 EDT]      0B solr-cores/
NuMacOS:Desktop mbagnall$ ./mc ls fsis
[2019-06-02 09:13:42 EDT]      0B crawler-data/
[2019-06-02 07:53:12 EDT]      0B databases/
[2019-06-02 09:00:40 EDT]      0B drupal-data/
[2019-06-02 07:52:31 EDT]      0B drupal-files/
[2019-06-02 07:52:09 EDT]      0B private-files/
[2019-06-02 09:09:46 EDT]      0B solr-cores/
NuMacOS:Desktop mbagnall$ ./mc ls fsis
mc: <ERROR> Unable to list folder. The specified bucket does not exist.
NuMacOS:Desktop mbagnall$ ./mc ls fsis
mc: <ERROR> Unable to list folder. The specified bucket does not exist.


And now that command does not seem to work at all.

**Comment by ElusiveMind** [comment]
Debug Info. I removed the x-Amz headers

NuMacOS:Desktop mbagnall$ mc --debug ls fsis
mc: <DEBUG> GET / HTTP/1.1
Host: fsis.itcon-files.com
User-Agent: MinIO (darwin; amd64) minio-go/v6.0.26 mc/2019-05-29T21:21:34Z
Authorization: AWS4-HMAC-SHA256 Credential=minio_fsis/20190602/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 404 Not Found
Connection: keep-alive
Date: Sun, 02 Jun 2019 13:34:48 GMT
Server: nginx
X-Served-By: Namecheap URL Forward
Content-Length: 0

mc: <DEBUG> Response Time:  168.683805ms

mc: <DEBUG> GET / HTTP/1.1
Host: fsis.itcon-files.com
User-Agent: MinIO (darwin; amd64) minio-go/v6.0.26 mc/2019-05-29T21:21:34Z
Authorization: AWS4-HMAC-SHA256 Credential=minio_fsis/20190602/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=**REDACTED**
Accept-Encoding: gzip

mc: <DEBUG> HTTP/1.1 404 Not Found
Connection: keep-alive
Date: Sun, 02 Jun 2019 13:34:48 GMT
Server: nginx
X-Served-By: Namecheap URL Forward
Content-Length: 0

mc: <DEBUG> Response Time:  77.467393ms

mc: <ERROR> Unable to list folder. The specified bucket does not exist.
 (1) ls.go:135 cmd.doList(..) Tags: [http://fsis.itcon-files.com/]
 (0) client-s3.go:1791 cmd.(*s3Client).listInRoutine(..)
 Release-Tag:RELEASE.2019-05-29T21-21-34Z | Commit:f4bb0b8850cb | Host:NuMacOS.fios-router.home | OS:darwin | Arch:amd64 | Lang:go1.12.4 | Mem:1.7 MB/72 MB | Heap:1.7 MB/66 MB

**Comment by kannappanr** [comment]
@ElusiveMind How was `MinIO` started? Is it running in `fs`, `stand alone erasure coded` mode or `distributed erasure coded mode` ? Anything else specific for us to reproduce the issue will be great.

**Comment by ElusiveMind** [comment]
I am not sure. I am running it through docker:

https://hub.docker.com/r/minio/minio/

**Comment by kannappanr** [comment]
can you type `mc admin info myminio` where `myminio` is the alias name to your minio deployment and paste the output here. You can redact the IPAddress if you want to.

**Comment by ElusiveMind** [comment]
This can't be good.

NuMacOS:~ mbagnall$ mc admin info fsis
mc: <ERROR> Cannot get service status. Failed to parse server response.

**Comment by harshavardhana** [comment]
> mc: Cannot get service status. Failed to parse server response.

This shows that your server and client are running too old versions and are incompatible. Update `mc` as well as you update your server.

**Comment by ElusiveMind** [comment]
Ok - Well I think the answer is I need to build my own Docker container with the latest Minio. That is really annoying. I am using the official one and the latest MC and if what you are saying is true, then they are out of sync. Let me work on that and get back to you.

**Comment by ElusiveMind** [comment]
Ok - I just updated my Minio docker container

NuMacOS:~ mbagnall$ mc admin info fsis
●  127.0.0.1:9000
   Status : online
   Uptime : 20 seconds 
  Version : 2019-06-01T03:46:14Z
  Storage : Used 268 GiB

**Comment by kannappanr** [comment]
We looked at the `mc debug response`

```
mc: HTTP/1.1 404 Not Found
Connection: keep-alive
Date: Sun, 02 Jun 2019 13:34:48 GMT
Server: nginx
X-Served-By: Namecheap URL Forward
Content-Length: 0
```

The request is not reaching `MinIO` server as you can see that the `Server` header is set to `nginx`, in the case of `MinIO` it is set to something like `Server: MinIO/DEVELOPMENT.2019-06-03T06-10-39Z`

Do you know what this header `X-Served-By: Namecheap URL Forward` does?

One more observation is that you are hitting `Nginx` using `mc` and not `MinIO` directly. Can you please connect to `MinIO` directly, by specifying the `MinIO` endpoint information instead of the `Nginx` endpoint and run the tests using `mc` again?

**Comment by ElusiveMind** [comment]
For this particular use case, I cannot. I am hosting in Azure using my employer account and government clients, so they will not let me open non-standard web ports for transmission or receiving. So as such, I cannot open port 9000 on the host to do this test. Which is why I am proxying subdomains through nginx to different Minio hosts on the same server.

I am not sure about the NameCheap URL forward. I am using NameCheap as the registrar for the domain name, but an nslookup on the hostname points directly to the IP address of my VPS and not NameCheap. I do get that this could be an issue and I identified that. The perplexing part is that it fails every _other_ request and not every request. That is the unnerving part of this. However with mc, it seems to be even less reliable.

This is why I posted my Nginx configuration for the different vhost configurations. 

Simply put - I am trying to determine definitively if this is an Nginx problem and how to fix it - or if it is a bug in Minio. I am sure I am not the only developer trying to proxy Minio through Nginx. 

I appreciate the attention. I hope we can find a resolution. It's very perplexing.

**Comment by kannappanr** [comment]
@ElusiveMind The issue here is that it is not hitting `MinIO` when you see the errors in `mc`. So, it will have something to do with the `nginx` settings and/or combination of other things that you have mentioned here. It will helpful to reduce the number of variables and reproduce the issue to isolate the actual root cause of the problem. As of now it doesn't look like it is reaching `MinIO`, when it fails.

**Comment by ElusiveMind** [comment]
I have resolved the issue. NameCheap - by default - adds a redirect entry to their DNS that messes with the IP's of subdomains. As such, it was going in round-robin fashion to two different IPs. The problem resolved when I removed the bad entry and DNS propagated. Sorry for the noise, but hopefully this helps out someone else who may have the same problem down the road.
